### PR TITLE
Add dynamic launcher portal

### PR DESCRIPTION
**Pull request by mwleeds**
The corresponding x-d-p PR is here:
https://github.com/flatpak/xdg-desktop-portal/pull/696

### CODE REVIEW

File: libportal/dynamic-launcher.c
Comment (GeorgesStavracas) [inline libportal/dynamic-launcher.c:526]: Shouldn't it return a `GIcon*` instead?

File: libportal/dynamic-launcher.c
Comment (GeorgesStavracas) [inline libportal/dynamic-launcher.c:251]: Ditto

**Comment by GeorgesStavracas** [comment]
I don't have much else to say, this LGTM once we figure out the `GIcon*` situation

File: libportal/dynamic-launcher.c
Comment (mwleeds) [reply libportal/dynamic-launcher.c:526]: I guess it depends how closely we want the libportal API to match the underlying portal API, but I don't see much benefit. Using `GIcon` here and in the `xdp_portal_dynamic_launcher_prepare_install` signature means the caller doesn't have to do `g_icon_serialize`/`g_icon_deserialize`, but it also means libportal has to handle erroring out if the `GIcon` is not a `GBytesIcon` (or let the portal error out instead). Doesn't really seem like enough of a benefit to justify reworking this and the calling code?

Comment (mwleeds) [reply libportal/dynamic-launcher.c:526]: >it also means libportal has to handle erroring out if the GIcon is not a GBytesIcon (or let the portal error out instead)

I guess we can just replace `g_return_val_if_fail (g_variant_is_of_type (icon_v, G_VARIANT_TYPE ("(sv)")), NULL);` with `g_return_val_if_fail (G_IS_BYTES_ICON (icon), NULL);` when the `GIcon` is being passed in. But then we have to handle setting the error pointer if serialization fails (GLib doesn't set it for us). And similarly we would need to handle setting the error pointer if deserialization fails in `xdp_portal_dynamic_launcher_get_icon()`.

I don't see any use of `g_set_error()` in libportal code aside from in `get_under_snap()`, so I don't know if we want to start using it and make an error domain for libportal (or use something generic like `G_IO_ERROR_FAILED` but that's an anti-pattern). Seems like that would make it a bit harder to reason about when client code is checking error conditions, since the status quo is that errors are only set by portal implementations or at the layer of gdbus, I think.

Another reason it might not make sense to use `GIcon` is in case some clients would want the icon in serialized form when retrieving it with `xdp_portal_dynamic_launcher_get_icon()` or if some clients would already have it serialized and have to deserialize it to pass it as a `GIcon`.

Just some thoughts, not a strong opinion.

Comment (GeorgesStavracas) [reply libportal/dynamic-launcher.c:526]: If this code requires a specific implementation of GIcon, then perhaps it should use that as the accepted type. E.g. `GBytesIcon * xdp_portal_dynamic_launcher_... ()`. Though, on a second thought, using the implementation type directly would limit any future API changes.

Comment (mwleeds) [reply libportal/dynamic-launcher.c:526]: Perhaps we should stick with `GVariant` then?

Comment (GeorgesStavracas) [reply libportal/dynamic-launcher.c:520]: Let's go with GVariant for now

**Review by GeorgesStavracas — APPROVED** [review]
Please rebase, squash, and land it

**Pull request merged**
